### DISCUSSION TITLE
[Proposal] Syntactic sugar: allow "pseudo-types" for `is`-operator

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#807] [General]
@Unknown6656 commented on [Sun Jun 26 2016](https://github.com/dotnet/roslyn/issues/12212)

I would like to propose more "pseudo-types" for the `is`-operator, as I (and many other developers) often work with reflection and `System::Type` and have wondered, whether one could optimize code readabilty in such situations.

``` csharp
dynamic obj = "Hello, World!";

bool test0 = obj is string; //true (this already exists)

//new syntax proposals:
bool test1 = obj is class; //true, as the type is a managed class
bool test2 = obj is struct; //false, as the type is not a structure
bool test3 = obj is internal; //false, the type is publicly accessible
bool test4 = obj is public; //true
bool test5 = obj is <>; //false, as the type is not generic
bool test6 = obj is abstract; //false
bool test7 = obj is [Serializable]; //true, as the Serializable-Attribute is added to the type definition
bool test8 = obj is [ComVisible]; //true, as the ComVisible-Attribute is added to the type definition

//The previous two tests could be also combined as follows:
bool test9 = obj is [Serializable, ComVisible];

Assert.Equals(test7 & test8, test9); //true
```

As you can probably see, I would like to see different keywords/accessors/etc. being used with the `is`-operator, instead of using the circumbendibus syntax with with `System::Type`:

``` csharp
dynamic obj = "Hello, World!";
Type t = obj.GetType();

// etc.
bool test7 = t.IsSerializable;
bool test8 = t.GetCustomAttributes(typeof(ComVisibleAttribute), true).Count() > 0;
```

This proposal has multiple advantages:
- No need for CLR-changes, as this syntax can be translated by the compiler
- Much more readable and intuitive than the use of `System::Type`-instances<br/>
  (Especially for attribute and generic checks)
- Cleaner and far mor shorter code

<hr/>

One could propose the following syntax rules for the `is`-Operator:

```
expression := expression "is" pattern

pattern := typename // `typename` means the existing .NET type name pattern
        |  typestring
        |  "<" commata ">"
        |  accessmodifier
        |  "[" attributes "]"

commata := "," commata
        |  __emptystring__

typestring := "Enum" //this already exists
           |  "class"
           |  "struct"
           |  "interface"
           |  "Delegate" //also exists

accessmodifier := accessmodifier "," accessmodifier
               |  "public"
               |  "private"
               |  "internal"
               |  "protected"
               |  "abstract"
               |  "sealed"
               //NOTE: accessmodifiers must not repeat or contradict themselves (e.g.: `private,public`)

attributes := attributes "," attributes
           |  attributename //the .NET name of the attribute (the name suffix `Attribute` is optional)
```

_NOTE: The expression `bool test = obj is <,,>` would match any object instance `obj`, whoms type is a generic type with **exactly** three generic type arguments._

---

@HaloFour commented on [Sun Jun 26 2016](https://github.com/dotnet/roslyn/issues/12212#issuecomment-228620849)

What happens in the case that `obj` is `null`?  Do all of the above tests then fail?

Also, neither `obj is abstract` or `obj is interface` make much sense.  The implementation of `obj` can't be either.  Do you mean to test if `obj` happens to have an abstract base class, or if `obj` happens to implement any interface?  What possible use case could that have?

Speaking of use cases, when would it use useful to know that `obj` happens to be a generic class of some specific arity?  What can you do with that information?

---

@Unknown6656 commented on [Tue Jun 28 2016](https://github.com/dotnet/roslyn/issues/12212#issuecomment-229005755)

@HaloFour: Well, you are right, that `... is interface` or `... is abstract` do not make sense...+

It would be better to implement something like this to test for multiple interfaces:

``` csharp
bool res = "hello" is IConvertible, IEnumerable; // true, as system::string can be cast to both interfaces
```

Concerning your `null`-question: The result should always be `false` (as the `is`-comparator would return in previous C# versions)

Concerning your last question, whether it would be useful: I have often used generic type creation or generic type pattern matching in the past, and I think that it could be rather useful.
One could also use this feature inside various .dotnet projects (e.g. various compilers or analysers which require pattern matching/analyzation)...

---

@bondsbw commented on [Fri Jul 01 2016](https://github.com/dotnet/roslyn/issues/12212#issuecomment-230041700)

While we're at it, perhaps comparisons directly on the type would also be useful:

``` c#
void Foo<T>()
{
    //new syntax proposals:
    Console.WriteLine($"T is class: {T is class}");
    Console.WriteLine($"T is struct: {T is struct}");
    Console.WriteLine($"T is internal: {T is internal}");
    Console.WriteLine($"T is public: {T is public}");
    Console.WriteLine($"T is <>: {T is <>}");
    Console.WriteLine($"T is abstract: {T is abstract}");
    Console.WriteLine($"T is [Serializable]: {T is [Serializable]}");
    Console.WriteLine($"T is [ComVisible]: {T is [ComVisible]}");

    //The previous two tests could be also combined as follows:
    Console.WriteLine($"T is [Serializable, ComVisible]: {T is [Serializable, ComVisible]}");
}
```

``` c#
Foo<string>();
// T is class: true
// T is struct: false
// T is internal: false
// T is public: true
// T is <>: false
// T is abstract: false
// T is [Serializable]: true
// T is [ComVisible]: true
// T is [Serializable, ComVisible]: true

Foo<int>();
// T is class: false
// T is struct: true
// T is internal: false
// T is public: true
// T is <>: false
// T is abstract: false
// T is [Serializable]: true
// T is [ComVisible]: true
// T is [Serializable, ComVisible]: true
```

---

@bondsbw commented on [Fri Jul 01 2016](https://github.com/dotnet/roslyn/issues/12212#issuecomment-230041962)

Though now that I think about it, the syntax `T : class` might be more appropriate given similar usage for type parameters constraints.  I don't know.

---

@Unknown6656 commented on [Sat Jul 09 2016](https://github.com/dotnet/roslyn/issues/12212#issuecomment-231538261)

@bondsbw: Your last proposal could be an issue for readability in the following situation:

``` csharp
Type t = ....;
bool res = t : public ? t : abstract ? t : class : t : struct : t : [Serializable];
```

is identical to:

``` csharp
bool res = (t is public) ? ((t is abstract) ? (t is class) : (t is struct)) : (t is [Serializable]);
```

I think that using a colon (`:`) instead of `is` renders the code less readable (independent of bracket-usage)

---

@bondsbw commented on [Sat Jul 09 2016](https://github.com/dotnet/roslyn/issues/12212#issuecomment-231538964)

Good point.  I can't think of a reason to prefer the colon syntax which justifies the confusion caused by what you wrote.

**Comment by Opiumtm** [comment]
Could it solve #662 problem?

So, would it possible to write code like this?
```C#
public void GenericMethod<T>(T obj)
{
    if (obj is struct s)
    {
        OptimizedForStructs(s); 
        // compiler CERTAINLY KNOWS that s is struct
        // so it's certainly safe to invoke method with "struct" type constraint
    }
}

private void OptimizedForStructs<T>(T obj) where T : struct
{
    // optimized for value types code
}
```

**Comment by gafter** [comment]
@Opiumtm How would you imagine that would be translated into IL? Or into machine code?

**Comment by Opiumtm** [comment]
@gafter 
JIT already know everything about actual type of object on method invoke.

It's about "more restrictive" generic type argument of `OptimizedForStructs<T>(T obj) where T : struct`

Runtime would throw exception (on generic type argument check) if this method is invoked with object instance of reference type, but it would be never invoked with reference type instance, thanks to `obj is struct s` check.

**Comment by Opiumtm** [comment]
@gafter 
There are known methods to invoke method with generic argument from non-generic code or from less type-restricted code without reflection. Manual IL emit at runtime does allow to invoke generic method from non-generic method or less-type-restricted generic method. Why not do it during compilation (i.e. emit same IL code which emitted at runtime when developer try to overcome that generic method invoke restriction) when type safety checks can be done by compiler?

**Comment by gafter** [comment]
@Opiumtm We can't do it at compile-time because the compiler doesn't know *what* specific struct type the value will end up being at runtime. That will only be known at runtime.

**Comment by Opiumtm** [comment]
@gafter 
https://www.red-gate.com/simple-talk/blogs/subterranean-il-callvirt-and-generic-types/

>Lets take a step back. The only reason we have to box the value type instance is to generate a MethodTable pointer as part of the heap object that the virtual method call can use. But, if it’s a value type, we already know the correct table to use at compile time, as value types are all sealed. What if we could specify the MethodTable to use without having to box the instance?
This is where the constrained. prefix instruction comes in (note the dot at the end). This specifies which MethodTable the following callvirt should use if one isn’t already provided by the this pointer. It also modifies the callvirt to take a this of type & rather than O; if the constrained type is a reference type, the & is a pointer to an object reference (so it’s a void** in C parlance), if it’s a value type, it’s a pointer to a value type instance (for the same reasons that call on a value type takes a &).

**Comment by gafter** [comment]
> But, if it’s a value type, we already know the correct table to use at compile time

No, given the new language construct proposed here, knowing that it is a value type no longer implies knowing what that value type is. The compiler could generate a constrained call. But what value type should the compiler place into that constrained call?

**Comment by Opiumtm** [comment]
@gafter 
`private void OptimizedForStructs<T>(T obj) where T : struct` 
is **generic** method
So exact type of struct isn't known and shouldn't be known at compile time **as with any generic method invoke already**!

**Comment by iam3yal** [comment]
@Opiumtm 

> So exact type of struct isn't known and shouldn't be known at compile time as with any generic method invoke already!

The exact type is definitely known by the compiler, the part that is unknown is the instance of the type which is only known at run-time so how do you expect the compiler to verify that when it doesn't even know what `obj` points to?

**Comment by gafter** [comment]
> The exact type is definitely known by the compiler

@eyalsk No, in the (proposed) code
``` c#
static void M(object o)
{
    if (o is struct s) M(s);
}
```
where
``` c#
public static void M<T>(T s) where T : struct {}
```
The exact static type that is being passed to `M` is definitely not known by the compiler. If you think it is known by the compiler, what type do you think the compiler knows is being passed to `M`?

**Comment by iam3yal** [comment]
@gafter Okay, might be poor wording on my part but I really don't understand you because the compiler definitely knows what the non-generic M expects and this is what I was referring to, now, the part that isn't known is where `o` points to, might be an instance of a `struct` but this isn't known at compile-time.

Basically, what I'm trying to say is that the following part `if (o is struct s) M(s);` is only known at run-time and the compiler doesn't know anything about `o` except that it is an object.

**Comment by gafter** [comment]
In order to emit the invocation, the compiler needs to know the type.

**Comment by Opiumtm** [comment]
@gafter 
How about generics? Generics by definition doesn't have "known" type on invoke.

**Comment by iam3yal** [comment]
@Opiumtm But you still need to pass the type argument at the call site.

**Comment by Opiumtm** [comment]
@eyalsk 
And so what?
What's the problem to pass type argument?

```C#
private void OptimizedForStructs<T>(T obj) where T : struct
```
What's problem to pass `T` as type argument here?
It's known at compile time.

**Comment by iam3yal** [comment]
@Opiumtm 

> What's the problem to pass type argument?

Did I say that there's a problem? no, I did not! I said that when it comes to generics you have to specify the type argument so it's known at the time of invocation...

**Comment by jnm2** [comment]
@gafter Can you explain the problem? I'm not getting it.

**Comment by gafter** [comment]
@jnm2 Yes, the problem is that in the code that is [proposed to be supported](https://github.com/dotnet/csharplang/issues/807#issuecomment-322073520), the compiler does not know the type, and the programmer is not required to pass it anywhere. There is no way to generate IL for the proposed source code.

**Comment by jnm2** [comment]
Oh, got it.

**Comment by gafter** [comment]
Specifically, what type should `M(object o)` use for the type argument in the call to `M<T>(T s)`?

**Comment by Opiumtm** [comment]
@gafter
`System.ValueType` for structs (or any other value type), maybe?

**Comment by gafter** [comment]
@Opiumtm `System.ValueType` isn't a `struct` type, and that would require boxing of the argument when being passed, which is exactly the thing this code is presumably trying to avoid.

**Comment by Opiumtm** [comment]
@gafter 
After all, if you invoke generic method from non-generic context, you either pass explicit type (in such case all would work already in C#), or pass `object` which is already boxed in case of contained struct.

Then let's go back and look and code example I quoted

```C#
public void GenericMethod<T>(T obj)
{
    if (obj is struct s)
    {
        OptimizedForStructs(s); 
        // compiler CERTAINLY KNOWS that s is struct
        // so it's certainly safe to invoke method with "struct" type constraint
    }
}

private void OptimizedForStructs<T>(T obj) where T : struct
{
    // optimized for value types code
}
```

Generic method `OptimizedForStructs<T>(T obj) where T : struct` is invoked from generic context - `public void GenericMethod<T>(T obj)`
So, it **this** case you would pass `T` as type argument. The only difference here is constraints on type argument `T`

`OptimizedForStructs<T>(T obj) where T : struct` is constrained to structs only, and caller context method `public void GenericMethod<T>(T obj)` does not constrained (so `T` can be struct or reference type here)

When we check it here `if (obj is struct s)` - we ensure that `s` (remaining of type `T`) is always struct so it's safe to invoke constrainted to structs method `OptimizedForStructs<T>(T obj) where T : struct` as argument `obj` (of type `T`) should always be of value type and runtime exception on type argument check would never occur. So it's safe to do constrained `callivrt` with type `T`.

If you invoke same method from non-generic context, it could be `object` and if it's struct, it's obviously have been boxed already at non-generic context.

**Comment by jnm2** [comment]
> So, it this case you would pass T as type argument. The only difference here is constraints on type argument T

I'm pretty sure that's illegal IL. You can't use a method's generic type parameter as a type outside the method.

Just tried it. It throws a runtime exception:

> Common Language Runtime detected an invalid program.

**Comment by MgSam** [comment]
I don't get this proposal. Why is this use case so important that it should be elevated to the level of language syntax? How often are people actually writing this type of code? And even if you were writing this kind of code every single day, what's wrong with just using helper methods? And why is this little slice of reflection more deserving of new syntax than all the other use cases of reflection?

I wish the community would push back a little harder on all of the "I occasionally do this thing so lets make a dedicated syntax for it" type requests.

**Comment by jnm2** [comment]
> I wish the community would push back a little harder on all of the "I occasionally do this thing so lets make a dedicated syntax for it" type requests.

When the issue gets to this point I tend to agree, but there's nothing wrong with putting stuff out there to see if it's something everyone is interested in.

**Comment by HaloFour** [comment]
I can see some use cases for stuff like generic specialization within a method but I might approach it differently and I honestly don't know how often it would be that useful.

🍝  Looking at @Opiumtm 's specific scenario maybe some syntax like this?

```cs
public void GenericMethod<T>(T obj)
{
    if (T is struct)
    {
        OptimizedForStructs(s); 
        // compiler CERTAINLY KNOWS that T is a struct
    }
}

private void OptimizedForStructs<T>(T obj) where T : struct
{
    // optimized for value types code
}
```

The difference being the application of the `is` operator on the generic type argument directly and the operand would be any generic type constraint.  If the condition succeeds the compiler could then consider that generic type argument to match that constraint.

Another example:

```cs
public int Count<T, TE>(T enumerable) where T : IEnumerable<TE> {
    if (T is IList<TE>) return enumerable.Count;
    // do other count here
}
```

**Comment by bondsbw** [comment]
@HaloFour Agreed [(more examples)](https://github.com/dotnet/roslyn/issues/12212#issuecomment-230041700).

**Comment by iam3yal** [comment]
@MgSam But is it really occasionally? many frameworks use reflection all the time.. testing frameworks, DI frameworks, UI frameworks and whatnot, if the language can make reflection better then why not?

C++ committee make the experience for templates better all the time.. does every C++ programmer create templates? well, I guess they are trying to make it easier but still, I know it's really comparing apples and oranges but the point is a language feature not always have to lean towards the majority... 

p.s. I really see this more of a new pattern than actually a new syntax to the language.

**Comment by MgSam** [comment]
@eyalsk Exactly my point. This proposal covers what, 1% of reflection use cases? To cover them all would be a mammoth undertaking. And for what benefit? It's not like it gives you any additional compile time benefits (as `nameof` does). It's just yet another way of doing the exact same thing you can already do.

I write C# code generators reasonably frequently- the only one of these use cases I've ever actually done myself was check a type for an attribute. But it's simple enough to do with a helper method. And most developers *never* write that sort of code.

**Comment by iam3yal** [comment]
@MgSam Okay, I see your point and I tend to agree.

**Comment by Joe4evr** [comment]
> if the language can make [reflection] better then why not?

Because the language team doesn't want to promote reflection by giving it an easy syntax.

**Comment by iam3yal** [comment]
@Joe4evr 

> Because the language team doesn't want to promote reflection by giving it an easy syntax.

I don't think we should state what they want or think before having an official words on it and until then speak for ourselves. 

Besides, if they didn't want to discuss this then why @gafter bothered to copy/paste it to this repo? he could have just closed it.

In my opinion the proposal isn't bad and like I wrote before to me it's more of a pattern than a syntax, it does make things more readable, however, it's very niche and like @MgSam wrote it covers only tiny bit of reflection use cases, still, it might have some value.

**Comment by Joe4evr** [comment]
> I don't think we should state what they want or think before having an official words on it and until then speak for ourselves.

[Mads' own words.](https://github.com/dotnet/roslyn/issues/128#issuecomment-239907842)

**Comment by iam3yal** [comment]
@Joe4evr Fair enough but still, does this mean we shouldn't discuss the idea? :)

**Comment by benaadams** [comment]
The common runtime patterns for class/struct are:
```csharp
if (T is struct)
```
```csharp
if (default(T) != null)
```
And
```csharp
if (T is class)
```
```csharp
if (default(T) == null)
```
Nullable; is more tricky :)

**Comment by benaadams** [comment]
48 occurrences of `default(T) == null` in corefx and coreclr (including tests)
13 occurrences of `default(T) != null` in corefx and coreclr (including tests)

**Comment by benaadams** [comment]
Related but more limited in scope "Pattern match via generic constraint" https://github.com/dotnet/csharplang/issues/905

**Comment by Rekkonnect** [comment]
With the new list pattern syntax using `[ ... ]`, the attributes will have to be denoted in another way. Overall I'm still a big fan of many of those patterns. The accessibility modifier patterns don't touch me that much, but I fail to find reasons to reject them.
